### PR TITLE
CLOUDSTACK-8820: Support for VMware vCenter 6 data center

### PR DESCRIPTION
**Pull request by sureshanaparti**
CLOUDSTACK-8820: Showing error when try to add advance zone using VMware ESXi 6.0 host

Summary: In vCenter 6.0, response headers need to be fetched after service login for server cookie unlike previous versions of vCenter.

### CODE REVIEW

**Comment by bhaisaab** [comment]
Great PR @sureshanaparti I was going to work on supporting this as ESXi6 works well on top of KVM; I can test this for you next week; LGTM on code change (not tested though)

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java [hunk 1]
```diff
@@ -152,13 +152,26 @@ public void connect(String url, String userName, String password) throws Excepti
         @SuppressWarnings("unchecked")
         Map<String, List<String>> headers = (Map<String, List<String>>)((BindingProvider)vimPort).getResponseContext().get(MessageContext.HTTP_RESPONSE_HEADERS);
         List<String> cookies = headers.get("Set-cookie");
+
+        vimPort.login(serviceContent.getSessionManager(), userName, password, null);
+
+        if (cookies == null) {
+            @SuppressWarnings("unchecked")
```
Comment (DaanHoogland) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: why suppress? please add a comment

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java [hunk 2]
```diff
@@ -577,41 +590,48 @@ public ManagedObjectReference getDecendentMoRef(ManagedObjectReference root, Str
...
+                List<DynamicProperty> propary = oc.getPropSet();
+                if (type == null || type.equals(mor.getType())) {
+                    if (propary.size() > 0) {
+                        String propval = (String)propary.get(0).getVal();
+                        if (propval != null && name.equalsIgnoreCase(propval))
+                            return mor;
+                    }
                 }
             }
+        } catch (Exception e) {
```
Comment (DaanHoogland) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: please catch specific exceptions, not just anything

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java [hunk 3]
```diff
@@ -577,41 +590,48 @@ public ManagedObjectReference getDecendentMoRef(ManagedObjectReference root, Str
...
+            s_logger.debug("Failed to get mor for name: " + name + " and type: " + type, e);
```
Comment (DaanHoogland) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: (mor == more || mor is MoRef)?

**Comment by DaanHoogland** [comment]
change makes sense but I have some comments on the coding style.

**Comment by remibergsma** [comment]
@sureshanaparti Nice work! The PR title confused me a bit though. What about:
CLOUDSTACK-8820: Support for VMware vCenter 6 data center

@bhaisaab Would appreciate it if you could test this. Thanks! Nice feature to go with ACS 4.6 :-)

**Comment by bhaisaab** [comment]
@remibergsma sure, but can do this next week (lack infra right now)

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java (hunk 1 above)
Comment (sureshanaparti) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: to suppress the warnings for the unchecked cast from Object to Map<String,List<String>> on extracting the http response headers from the response context. This is also set on vmware sample programs.

**Comment by sureshanaparti** [comment]
@remibergsma That sounds good.

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java (hunk 2 above)
Comment (sureshanaparti) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: update the code to catch specific exceptions

File: vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java (hunk 3 above)
Comment (sureshanaparti) [inline vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java]: MoRef i.e VMware ManagedObjectReference. Updated the log.

**Comment by DaanHoogland** [comment]
FWIW LGTM

**Comment by sateesh-chodapuneedi** [comment]
@sureshanaparti code LGTM.
I will test this patch today in my lab and will get back.

**Comment by bhaisaab** [comment]
Alright, let's wait for lab tests to confirm the patch and only then we should merge it.

Thanks for helping test it @sateesh-chodapuneedi

**Comment by sateesh-chodapuneedi** [comment]
Tested adding ESXi 6.0 host to advanced ACS zone successfully.
Session logic is working fine. 

@bhaisaab you might want to go ahead merge it. Thanks.

**Pull request merged**